main.c

```diff
@@ -0,0 +1,412 @@
+        /***
+            On Linux the splice() and sendfile() syscalls can be useful for us (see
+            their descriptions taken from the respective man pages below), on other
+            operating systems there might be functions with the same names, but
+            potentially doing something completely different.
+
+            splice() moves  data  between two file descriptors without copying be‐
+            tween kernel address space and user address space.  It transfers up  to
+            len bytes of data from the file descriptor fd_in to the file descriptor
+            fd_out, where one of the file descriptors must refer to a pipe.
+
+            sendfile()  copies  data  between one file descriptor and another.  Be‐
+            cause this copying is done within the kernel, sendfile() is more  effi‐
+            cient than the combination of read(2) and write(2), which would require
+            transferring data to and from user space.
+            The   in_fd   argument   must  correspond  to  a  file  which  supports
+            mmap(2)-like operations (i.e., it cannot be a socket or a pipe).
+        ***/
+        ssize_t (*sendfile_fn)(int out_fd, int in_fd, off_t *offset, size_t count);
+        if (S_ISFIFO(in_fd_stat.st_mode)) {
+            sendfile_fn = splice_sendfile;
+        } else {
+            sendfile_fn = sendfile;
+        }
+
+        if (fsync_interval == 0) {
+            fsync_interval = len;
+        }
+        ssize_t ret;
+        off_t n_unsynced = 0;
+        do {
+            ret = sendfile_fn(out_fd, in_fd, 0, MIN_SIZE(len, fsync_interval));
+            if (ret > 0) {
+                len -= ret;
+                stats.total_bytes += ret;
+                n_unsynced += ret;
+                if (n_unsynced >= fsync_interval) {
+                    fsync(out_fd);
+                    n_unsynced = 0;
+                }
+            }
+        } while ((ret > 0) && (len > 0));
+        success = ((ret == 0) || ((ret > 0) && (len == 0)));
+        error = errno;
+    }
+#endif  /* __linux__ */
+
+    close(in_fd);
+    close(out_fd);
+
+    if (!success) {
+        if (error != 0) {
+            fprintf(stderr, "Failed to copy data: %s\n", strerror(error));
+            printf("Total bytes written: %ju\n", (intmax_t) stats.total_bytes);
+        } else {
+            fprintf(stderr, "Failed to copy data\n");
+            printf("Total bytes written: %ju\n", (intmax_t) stats.total_bytes);
+        }
+        return EXIT_FAILURE;
+    } else {
+        if (write_optimized) {
+            puts("================ STATISTICS ================");
+            printf("Blocks written: %10zu\n", stats.blocks_written);
+            printf("Blocks omitted: %10zu\n", stats.blocks_omitted);
+            printf("Bytes written: %11ju\n", (intmax_t) stats.bytes_written);
+            printf("Total bytes: %13ju\n", (intmax_t) stats.total_bytes);
+            puts("============================================");
```

Comment:
this will be scrambled on smaller terminals (I am carrying about Craig here) and on larger ones can look strange. Either use the padding with printf, get the term sizes, or dont print fancy ASCI graphics. (just an advice)